src/main/java/org/purpurmc/purpurextras/modules/SleepPercentageMessageModule.java

```diff
@@ -0,0 +1,74 @@
+        PurpurConfig config = PurpurExtras.getPurpurConfig();
+        this.playerSleepMessage = config.getString("settings.chat.send-sleep-percentage-message.player-sleeping", "<grey><playername> has fallen asleep. <sleeping> out of <needed> required players in <worldname> are sleeping.");
+        this.nightSkipMessage = config.getString("settings.chat.send-sleep-percentage-message.skipping-night", "<grey>Enough players have slept! Skipping through the night in <worldname>.");
+    }
+
+    @Override
+    public void enable() {
+        PurpurExtras plugin = PurpurExtras.getInstance();
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @Override
+    public boolean shouldEnable() {
+        DefaultPermissions.registerPermission(sleepMessageBypass, "Allows player to not display a message in chat when they sleep", PermissionDefault.OP);
+        if ((playerSleepMessage == null || playerSleepMessage.isBlank()) && (nightSkipMessage == null || nightSkipMessage.isBlank()))
+            return false;
+        return PurpurExtras.getPurpurConfig().getBoolean("settings.chat.send-sleep-percentage-message.enabled", false);
+    }
+
+    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
+    public void onPlayerDeepSleep(PlayerDeepSleepEvent event) {
+        if (playerSleepMessage == null || playerSleepMessage.isBlank()) return;
+        if (event.getPlayer().hasPermission(sleepMessageBypass)) return;
+        World world = event.getPlayer().getWorld();
+        String playerName = event.getPlayer().getName();
+        String worldName = world.getName();
+        int currentSleepCount = 0;
+        int worldOnlineTotal = world.getPlayerCount();
+        Integer worldSleepPercent = world.getGameRuleValue(GameRule.PLAYERS_SLEEPING_PERCENTAGE);
+        Integer neededSleepers = (int) Math.ceil((worldSleepPercent / 100.0) * worldOnlineTotal);
+        List<Player> playerList = world.getPlayers();
+        for (Player player : playerList) {
+            if (player.isDeeplySleeping()) currentSleepCount += 1;
+        }
+        world.sendMessage(MiniMessage.miniMessage().deserialize(playerSleepMessage,
+                Placeholder.unparsed("playername", playerName),
+                Placeholder.unparsed("sleeping", String.valueOf(currentSleepCount)),
+                Placeholder.unparsed("needed", String.valueOf(neededSleepers)),
+                Placeholder.unparsed("worldname", worldName)));
+    }
+
+    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
+    public void nightSkip(TimeSkipEvent event) {
+        if (!event.getSkipReason().equals(TimeSkipEvent.SkipReason.NIGHT_SKIP)) return;
+        if (nightSkipMessage == null || nightSkipMessage.isBlank()) return;
+        String worldName = event.getWorld().getName();
+        event.getWorld().sendMessage(MiniMessage.miniMessage().deserialize(nightSkipMessage, Placeholder.unparsed("worldname", worldName)));
```

Comment:
PE has its own minimessage instance (PurpurExtras.getInstance().minimessage) in case we need to globally change something. Use that instead of calling the default one